### PR TITLE
Add feature to inject documentation annotations as *Restrictions into Metadata

### PR DESCRIPTION
**Pull request by georgend**
***Blocking Bugs***
Initial implementation blocked by a bug where APiDoctor mishandles YAML Front Matter, thereby misunderstanding the whole document model and breaking downstream csdl handling logic. 

***Features***
Extract Operation titles and descriptions and inject them into generated csdl as `*Restrictions` (Read, ReadByKey).
solution
For  the example above:-

- Extract corresponding page title as `Description`
- Extract corresponding page first paragraph as `LongDescription`

These injected `*Restrictions` will be extracted by [OpenAPI.NET.Odata](https://github.com/microsoft/OpenAPI.NET.OData/pull/206) and inserted into generated OpenAPI files as summaries and descriptions.
image

These summaries and descriptions will then flow into the Microsoft Graph PowerShell SDK Docs which extract the summaries and descriptions and insert them into the generated docs. 

Related to https://github.com/microsoft/OpenAPI.NET.OData/pull/206
Related to https://github.com/microsoftgraph/msgraph-sdk-powershell/issues/876

### CODE REVIEW

File: ApiDoctor.Publishing/CSDL/csdlwriter.cs [hunk 1]
```diff
@@ -1157,7 +1291,141 @@ private void BuildEntityContainer(EntityFramework edmx, string[] baseUrlsToRemov
...
+                            else
+                            {
+                                currentEntitySet.SourceMethods = methodCollection;
+                            }
+                        }
+                    }
+                    else if (SingletonPathRegEx.IsMatch(path))
+                    {
+                        var allSections = FullSingletonPathRegEx.Matches(path);
+                        foreach (var section in allSections)
```
Comment (BarryShehadeh) [inline ApiDoctor.Publishing/CSDL/csdlwriter.cs:1342]: Hey @georgend, I'm guessing this PR is not ready, but I'm just taking a peek. Let me know when it's ready. Also it would be cool if we can add a description to the change in the API Doctor versioning wiki.
An example would be nice.

File: ApiDoctor.Publishing/CSDL/csdlextensionmethods.cs [hunk 2]
```diff
@@ -39,7 +40,7 @@ namespace ApiDoctor.Publishing.CSDL
 
     internal static class CsdlExtensionMethods
     {
-
+        private static readonly Regex GuidRegex = new(@"[0-9a-fA-F\-]{32,36}", RegexOptions.Compiled);
```
Comment (baywet) [inline ApiDoctor.Publishing/CSDL/csdlextensionmethods.cs:43]:
```suggestion
        private static readonly Regex GuidRegex = new(@"[0-9a-f\-]{32,36}", RegexOptions.Compiled | RegexOptions.IgnoreCase);
```

File: ApiDoctor.Publishing/CSDL/csdlextensionmethods.cs [hunk 3]
```diff
@@ -189,8 +191,33 @@ internal static EntityFramework MergeWith(this EntityFramework framework1, Entit
...
+            string name)
+        {
+            var currentSet = edmx.GetISet(name, selector);
+            if (currentSet != null)
 
-
-
+                if (currentSet.SourceMethods != null)
+                {
+                    var sourceMethods = (MethodCollection)currentSet.SourceMethods;
```
Comment (baywet) [inline ApiDoctor.Publishing/CSDL/csdlextensionmethods.cs:213]:
```suggestion
                if (currentSet.SourceMethods is MethodCollection sourceMethods)
                {
```

**Review by baywet — COMMENTED** [review]
sorry, started reviewing but I won't have time to review fully before I go on vacations and I don't want to block it. left a few comments.

File: ApiDoctor.Publishing/CSDL/csdlwriter.cs [hunk 4]
```diff
@@ -610,7 +633,50 @@ private static Record GenerateHttpRequestMethodAnnotation(MethodDefinition metho
 
             return record;
         }
+        private static void FixUpAnnotations(string fullName, IODataAnnotatable annotatable, Dictionary<string, Annotations> schemaLevelAnnotations)
+        {
+            if (annotatable.Annotation?.Count > 0)
+            {
+                Annotations annotations;
+                if (schemaLevelAnnotations.TryGetValue(fullName, out annotations))
```
Comment (andrueastman) [inline ApiDoctor.Publishing/CSDL/csdlwriter.cs:641]:
```suggestion
                if (schemaLevelAnnotations.TryGetValue(fullName, out Annotations annotations))
```

File: ApiDoctor.Publishing/CSDL/csdlextensionmethods.cs [hunk 5]
```diff
@@ -189,8 +191,32 @@ internal static EntityFramework MergeWith(this EntityFramework framework1, Entit
 
         }
 
+        private static ISet GetISet(this EntityFramework edmx, 
+            string name, 
+            Func<EntityContainer, IEnumerable<ISet>> funcSelector)
+        {
+            var currentEntitySet = edmx.DataServices.Schemas.SelectMany(c => c.EntityContainers)
+                .SelectMany(funcSelector)
+                .FirstOrDefault(c => c.Name == name);
```
Comment (andrueastman) [inline ApiDoctor.Publishing/CSDL/csdlextensionmethods.cs:200]: QQ: Do we care if the comparison is case insensitive or not?

File: ApiDoctor.Validation/MethodDefinition.cs [hunk 6]
```diff
@@ -127,6 +128,11 @@ public static MethodDefinition FromRequest(string request, CodeBlockAnnotation a
         /// </summary>
         public string Request {get; private set;}
 
+        /// <summary>
+        /// The parsed Request data from the documentation. 
+        /// </summary>
+        public HttpRequest HttpRequest { get; private set; }
```
Comment (andrueastman) [inline ApiDoctor.Validation/MethodDefinition.cs:134]: I think we can undo this change. It is only set in line 107 but never really used elsewhere.

File: ApiDoctor.Publishing/CSDL/csdlwriter.cs [hunk 7]
```diff
@@ -1066,9 +1132,10 @@ private static ODataTargetInfo ParseRequestTargetType(string requestPath, Method
         }
 
         // EntitySet is something in the format of /name/{var}
-        private readonly static System.Text.RegularExpressions.Regex EntitySetPathRegEx = new System.Text.RegularExpressions.Regex(@"^\/(\w*)\/{var}$");
-        // Singleton is something in the format of /name
-        private readonly static System.Text.RegularExpressions.Regex SingletonPathRegEx = new System.Text.RegularExpressions.Regex(@"^\/(\w*)$");
+        private static readonly System.Text.RegularExpressions.Regex EntitySetPathRegEx = new (@"\/(\w*)\/{var}$", System.Text.RegularExpressions.RegexOptions.Compiled);
```
Comment (andrueastman) [inline ApiDoctor.Publishing/CSDL/csdlwriter.cs:1135]: 👍🏼

File: ApiDoctor.Publishing/CSDL/csdlwriter.cs [hunk 8]
```diff
@@ -1552,6 +1619,195 @@ public int GetHashCode(EnumerationDefinition obj)
...
+        }
+
+        private static void AddSingletonSourceMethods(EntityFramework edmx, MethodCollection methodCollection, string path)
+        {
+            if (SingletonPathRegEx.IsMatch(path))
+            {
+                var matches = FullSingletonPathRegEx.Matches(path);
+                var name = matches[0].Groups[1].Value;
+                edmx.AddSetSourceMethods(methodCollection, x => x.Singletons, name);
+                edmx.AddSetSourceMethods(methodCollection, x => x.EntitySets, name);
```
Comment (andrueastman) [inline ApiDoctor.Publishing/CSDL/csdlwriter.cs:1639]: Do we also need to the singletons to the `EntitySets` as well?

**Review by andrueastman — COMMENTED** [review]
Will just do some final testing to verify before merging this in.

**Review by andrueastman — APPROVED** [review]
